File: desktop/src/utils/slotMath.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { next, firstFree } from './slotMath.js';

test('next rueckt innerhalb der Seite vor', () => {
  assert.deepEqual(next(3, 4, 9), { page: 3, slot: 5 });
});

test('next blaettert am Seitenende um', () => {
  assert.deepEqual(next(3, 9, 9), { page: 4, slot: 1 });
  assert.deepEqual(next(3, 4, 4), { page: 4, slot: 1 });
  assert.deepEqual(next(3, 12, 12), { page: 4, slot: 1 });
});

test('firstFree bei leerem Binder ist Seite 1 Fach 1', () => {
  assert.deepEqual(firstFree([], 9), { page: 1, slot: 1 });
});

test('firstFree findet die Luecke, nicht das Ende', () => {
  // Fach 2 auf Seite 1 ist frei -- der Vorschlag muss dorthin, nicht hinter das letzte.
  const belegt = [{ page: 1, slot: 1 }, { page: 1, slot: 3 }, { page: 1, slot: 4 }];
  assert.deepEqual(firstFree(belegt, 9), { page: 1, slot: 2 });
});

test('firstFree geht bei voller Seite auf die naechste', () => {
  const volleSeite = [1, 2, 3, 4].map((slot) => ({ page: 1, slot }));
  assert.deepEqual(firstFree(volleSeite, 4), { page: 2, slot: 1 });
});

test('firstFree ueberspringt eine ganz volle Seite und findet die Luecke dahinter', () => {
  const belegt = [1, 2, 3, 4].map((slot) => ({ page: 1, slot }))
    .concat([{ page: 2, slot: 1 }, { page: 2, slot: 3 }]);
  assert.deepEqual(firstFree(belegt, 4), { page: 2, slot: 2 });
});

test('firstFree ignoriert Faecher jenseits der Seitengroesse', () => {
  // Ein Datensatz aus einer frueheren, groesseren Seitengroesse darf nicht dazu fuehren,
  // dass ein gueltiges Fach als belegt gilt.
  const belegt = [{ page: 1, slot: 1 }, { page: 1, slot: 7 }];
  assert.deepEqual(firstFree(belegt, 4), { page: 1, slot: 2 });
});

test('unsinnige Eingaben werfen nicht', () => {
  assert.doesNotThrow(() => next(0, 0, 0));
  assert.doesNotThrow(() => next(-3, -1, 9));
  assert.doesNotThrow(() => firstFree([{ page: 0, slot: 0 }], 0));
});

test('firstFree wirft auch bei sehr vielen belegten Faechern nicht (kein Math.max(...arr))', () => {
  // Zwillings-Kreuzprobe: Math.max(...array) sprengt ab ein paar zehntausend Elementen die
  // Aufrufstapel-Grenze der Engine -- ein Wurf, den es laut Kopfkommentar nie geben darf und
  // den Kotlins maxOf (kein Argument-Spread) so nicht kennt.
  const volleSeiten = [];
  for (let pg = 1; pg <= 150000; pg++) volleSeiten.push({ page: pg, slot: 1 });
  assert.doesNotThrow(() => firstFree(volleSeiten, 9));
  assert.deepEqual(firstFree(volleSeiten, 9), { page: 1, slot: 2 });
});

File: android/app/src/test/java/com/example/yugiohscanner/SlotMathTest.kt
package com.example.yugiohscanner

import com.example.yugiohscanner.ml.SlotMath
import org.junit.Assert.assertEquals
import org.junit.Test

class SlotMathTest {

    @Test fun `next rueckt innerhalb der Seite vor`() {
        assertEquals(3 to 5, SlotMath.next(3, 4, 9))
    }

    @Test fun `next blaettert am Seitenende um`() {
        assertEquals(4 to 1, SlotMath.next(3, 9, 9))
        assertEquals(4 to 1, SlotMath.next(3, 4, 4))
        assertEquals(4 to 1, SlotMath.next(3, 12, 12))
    }

    @Test fun `firstFree bei leerem Binder ist Seite 1 Fach 1`() {
        assertEquals(1 to 1, SlotMath.firstFree(emptySet(), 9))
    }

    @Test fun `firstFree findet die Luecke, nicht das Ende`() {
        // Fach 2 auf Seite 1 ist frei -- der Vorschlag muss dorthin, nicht hinter das letzte.
        val belegt = setOf(1 to 1, 1 to 3, 1 to 4)
        assertEquals(1 to 2, SlotMath.firstFree(belegt, 9))
    }

    @Test fun `firstFree geht bei voller Seite auf die naechste`() {
        val volleSeite = (1..4).map { 1 to it }.toSet()
        assertEquals(2 to 1, SlotMath.firstFree(volleSeite, 4))
    }

    @Test fun `firstFree ueberspringt eine ganz volle Seite und findet die Luecke dahinter`() {
        val belegt = (1..4).map { 1 to it }.toSet() + setOf(2 to 1, 2 to 3)
        assertEquals(2 to 2, SlotMath.firstFree(belegt, 4))
    }

    @Test fun `firstFree ignoriert Faecher jenseits der Seitengroesse`() {
        // Ein Datensatz aus einer frueheren, groesseren Seitengroesse darf nicht dazu fuehren,
        // dass ein gueltiges Fach als belegt gilt.
        val belegt = setOf(1 to 1, 1 to 7)
        assertEquals(1 to 2, SlotMath.firstFree(belegt, 4))
    }

    @Test fun `unsinnige Eingaben werfen nicht`() {
        SlotMath.next(0, 0, 0)
        SlotMath.next(-3, -1, 9)
        SlotMath.firstFree(setOf(0 to 0), 0)
    }

    @Test fun `maxOccupiedPage ist null ohne belegte Seite`() {
        assertEquals(null, SlotMath.maxOccupiedPage(emptyList()))
        assertEquals(null, SlotMath.maxOccupiedPage(listOf(null, null)))
    }

    @Test fun `maxOccupiedPage findet die hoechste belegte Seite`() {
        assertEquals(3, SlotMath.maxOccupiedPage(listOf(1, null, 3, 2)))
    }
}
